cmd/containerd-shim-runhcs-v1/task_hcs.go

```diff
@@ -1008,3 +1026,73 @@ func (ht *hcsTask) requestUpdateContainer(ctx context.Context, resourcePath stri
 	}
 	return ht.c.Modify(ctx, modification)
 }
+
+// publishNotifications monitors the containers notification queue.
+// Currently only OOM events are published.
+//
+// This MUST be called via a goroutine to wait on a background thread.
+// Returned error is used for scoping, and not meant to be consumed
+func (ht *hcsTask) publishNotifications(ctx context.Context) {
+	ctx, span := trace.StartSpan(ctx, "hcsTask::publishNotifications")
+	defer span.End()
+	var err error
+	defer func() { oc.SetSpanStatus(span, err) }()
+	span.AddAttributes(trace.StringAttribute(logfields.TaskID, ht.id))
+	entity := log.G(ctx).WithField(logfields.TaskID, ht.id)
+
+	var nq *queue.MessageQueue
+	nq, err = ht.c.Notifications()
+	if err != nil {
+		// includes `notifications.ErrNotificationsNotSupported`, where container
+		// does not publish any notifications
+		entity.WithError(err).Warning("unable to retrieve container notifications channel")
+		return
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			var v interface{}
+			v, err = nq.ReadOrWait()
+			if err != nil {
+				if err == queue.ErrQueueClosed {
+					// set err to nil for oc.SetSpanStatus
+					err = nil
```

Comment:
Maybe it would be better to just call this "queueErr" or something and that way we don't need to worry about setting the err to nil for `SetSpanStatus`

Reply:
I would have to set `err = queueErr` on the else branch then right? If I did want non-`ErrQueueClosed` errors to be set for the span status?

Reply:
Agh, this is another reason why I like named return variables more 😅